core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts

```diff
@@ -0,0 +1,412 @@
+
+const topicsTab = 'a.e2e-test-topics-tab';
+const topic = 'a.e2e-test-topic-name';
+const addTopicButton = 'button.e2e-test-create-topic-button';
+const topicNameField = 'input.e2e-test-new-topic-name-field';
+const topicUrlFragmentField = 'input.e2e-test-new-topic-url-fragment-field';
+const topicWebFragmentField = 'input.e2e-test-new-page-title-fragm-field';
+const topicDescriptionField = 'textarea.e2e-test-new-topic-description-field';
+const createTopicButton = 'button.e2e-test-confirm-topic-creation-button';
+const saveTopicButton = 'button.e2e-test-save-topic-button';
+
+const addSubTopicButton = 'button.e2e-test-add-subtopic-button';
+const subTopicTitleField = 'input.e2e-test-new-subtopic-title-field';
+const subTopicUrlFragmentField =
+  'input.e2e-test-new-subtopic-url-fragment-field';
+const subTopicDescriptionEditorToggle = 'div.e2e-test-show-schema-editor';
+const createSubTopicButton = '.e2e-test-confirm-subtopic-creation-button';
+
+const skillDescriptionField = 'input.e2e-test-new-skill-description-field';
+const skillReviewMaterialHeader = 'div.e2e-test-open-concept-card';
+const addSkillButton = 'button.e2e-test-add-skill-button';
+const confirmSkillCreationButton =
+  'button.e2e-test-confirm-skill-creation-button';
+
+const addStoryButton = 'button.e2e-test-create-story-button';
+const storyTitleField = 'input.e2e-test-new-story-title-field';
+const storyDescriptionField = 'textarea.e2e-test-new-story-description-field';
+const storyUrlFragmentField = 'input.e2e-test-new-story-url-fragment-field';
+const createStoryButton = 'button.e2e-test-confirm-story-creation-button';
+const saveStoryButton = 'button.e2e-test-save-story-button';
+const saveStoryMessageInput = 'textarea.e2e-test-commit-message-input';
+const publishStoryButton = 'button.e2e-test-publish-story-button';
+
+const addChapterButton = 'button.e2e-test-add-chapter-button';
+const chapterTitleField = 'input.e2e-test-new-chapter-title-field';
+const chapterExplorationIdField = 'input.e2e-test-chapter-exploration-input';
+const createChapterButton = 'button.e2e-test-confirm-chapter-creation-button';
+
+export class CurriculumAdmin extends BaseUser {
+  /**
+   * Function for navigating to the topic and skills dashboard page.
+   */
+  async navigateToTopicAndSkillsDashboardPage(): Promise<void> {
+    await this.goto(topicAndSkillsDashboardUrl);
+  }
+
+  /**
+   * Function for creating a skill in the topics and skills dashboard.
+   */
+  async createSkill(): Promise<void> {
+    await this.openTopicEditor();
+    await this.clickOn(addSkillButton);
+    await this.type(skillDescriptionField, 'Test Skill 1');
+    await this.clickOn(skillReviewMaterialHeader);
+    await this.clickOn(richTextAreaField);
+    await this.type(
+      richTextAreaField,
+      'This is a test skill for curriculum admin.'
+    );
+    await this.page.waitForSelector(
+      `${confirmSkillCreationButton}:not([disabled])`
+    );
+    await this.page.waitForTimeout(500);
```

Comment:
Ideally we shouldn't be waiting a hard-coded number of seconds to pass as it makes the test more flaky. Either the test doesn't wait long enough for the necessary UI components to load into the correct state which will cause the test to fail, or the test waits much longer than expected and passes. The same test passes on some test runs, while it fails on other test runs, which is what I mean by "flaky tests".

For each `waitForTimeout` statement, consider investigating why the test would fail without the statement. The other `waitFor...` methods such as `waitForSelector` with options `visible` or `hidden` are often much more reliable and it clarifies what state the surrounding UI needs to be in before executing the user action.

Ditto for all the other `waitForTimeout` statements you've added.

Reply:
The tests fail in the absence of the timeouts because of modal transitions - when modals open or close. That leads to the click never happening, which is why the timeout is needed and waitFor methods don't suffice. I've only left in the ones that are absolutely necessary as far as I'm aware.

If I'm not wrong, the transitions were of 0.3 seconds so 0.5 seconds is a safe space for the tests to pass in my opinion.

Reply:
I seem to recall @StephenYu2018 has some code in one of his PRs (see withinModal() in puppeteer-utils.js [here](https://github.com/oppia/oppia/pull/19869/files#diff-94cb0186ac035dfd0c36611e18f5832cd2c04e097de358044814f7e0cf13371fR217)) that introduces a general "do stuff in modal" functionality. Maybe we can reuse that idea here, if this only happens in modals?

Reply:
@seanlip I still need to merge the PR before the rest of the acceptance test team can access `withinModal()`.

@Vir-8 For now, try something like `waitForSelector(selector, {visible: true})` to wait for the modal to appear before executing user interactions with the modal. Also you could try `waitForSelector(selector, {hidden: true})` to wait for the modal to completely disappear before executing user interactions with UI components behind the modal. Also, these techniques aren't just for modals; you can apply them to any UI element that appears and/or disappears upon user interactions.

Reply:
@seanlip @StephenYu2018 Thanks for the suggestions, appreciate it!